sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java

```diff
@@ -396,6 +398,44 @@ public BoundedWindow window() {
 
 
   /**
+   * Annotation for declaring and dereferencing state cells. <i>Not currently supported by any
+   * runner</i>.
+   *
+   * <p>To declare a state cell, create a field of type a {@link StateSpec} and annotate it
+   * with a {@link StateId} like so:
+   *
+   * <code>
+   *   new DoFn<KV<Key, Foo>, Baz>() {
+   *     @StateId("my-state-id")
+   *     private final StateSpec<K, ValueState<MyState>> myStateSpec =
+   *         StateSpecs.value(new MyStateCoder());
+   *   }
+   * </code>
+   *
+   * <p>To use a declared state cell during processing, add a parameter of the appropriate
```

Comment:
Added some bullet points. I think the javadoc will evolve, despite worries to the contrary, because the validation that we choose to enforce will evolve too.

`StateSpec` fields without a `StateId` are benign, since they can't be used. We can add a short-circuit failure with a nice message at some point, but I'd rather not add arbitrary restrictions yet.